### PR TITLE
[Mobile]add knobs to set QUIC connection options and client connection options

### PR DESCRIPTION
**Pull request by RenjieTang**
Commit Message: add knobs to set QUIC connection options and client connection options
Additional Description: This allows Envoy Mobile applications to set Quiche's connection options so that more performance tuning can be done.
Risk Level: Low
Testing: Unit tests
Docs Changes: n/a
Release Notes: n/a
Platform Specific Features: Mobile only

### CODE REVIEW

File: configs/BUILD [hunk 1]
```diff
@@ -43,8 +43,8 @@ filegroup(
         "//bazel:disable_admin_functionality": [],
         "//conditions:default": [
             "envoy-demo.yaml",
-            "freebind/freebind.yaml",
             "envoy-tap-config.yaml",
+            "freebind/freebind.yaml",
```
Comment (RyanTheOptimist) [inline configs/BUILD:47]: Is this related? Looks like a few BUILD files are being reformattted.

File: api/envoy/config/core/v3/protocol.proto [hunk 2]
```diff
@@ -94,6 +94,12 @@ message QuicProtocolOptions {
   // Probes the peer at the configured interval to solicit traffic, i.e. ACK or PATH_RESPONSE, from the peer to push back connection idle timeout.
   // If absent, use the default keepalive behavior of which a client connection sends PINGs every 15s, and a server connection doesn't do anything.
   QuicKeepAliveSettings connection_keepalive = 5;
+
+  // An array of QUIC connection options represented in a serialized string, separated by commas.
+  string connection_options = 6;
```
Comment (RyanTheOptimist) [inline api/envoy/config/core/v3/protocol.proto:99]: Since this is a proto, I wonder if this should be a repeated string, one per option. That would see a bit more idiomatic than a text-delimited string.

File: mobile/library/cc/engine_builder.h [hunk 3]
```diff
@@ -152,6 +152,8 @@ class EngineBuilder {
   EngineBuilder& enableSocketTagging(bool socket_tagging_on);
 #ifdef ENVOY_ENABLE_QUIC
   EngineBuilder& enableHttp3(bool http3_on);
+  EngineBuilder& setHttp3ConnectionOptions(std::string options);
```
Comment (RyanTheOptimist) [inline mobile/library/cc/engine_builder.h:155]: I wonder if this should be a span<string> or if we prefer to use string to be simpler?

**Review by RyanTheOptimist — COMMENTED** [review]
Thanks for doing this! Small question about the format of the proto?

File: api/envoy/config/core/v3/protocol.proto (hunk 2 above)
Comment (RenjieTang) [reply api/envoy/config/core/v3/protocol.proto:99]: Great question! I thought about it too when I wrote this and I believe using a single string might be slightly better because of the jni layer. 
Here's a comparison of the 2 approaches:
**Use repeated strings**: The application calls addConnectionOption() multiple times->The options get stored in a Java array->The Java list gets converted to a jni 2d byte array via a for loop->The 2d byte array gets converted to a vector<string> via another for loop->The vector is iterated to populate the proto field->The protos are read into the actual QuicTagVector via a loop.
**Use a serialized string**: The application is responsible for serializing the copt string->The string gets passed all the way down to the string proto field->The string is deserialized by calling ParseQuicTagVector().

We slightly reduce the readability here so that we save computations on doing conversions.

File: mobile/library/cc/engine_builder.h (hunk 3 above)
Comment (RenjieTang) [reply mobile/library/cc/engine_builder.h:155]: ditto.
I believe having a string is simpler and faster

File: api/envoy/config/core/v3/protocol.proto (hunk 2 above)
Comment (RyanTheOptimist) [reply api/envoy/config/core/v3/protocol.proto:99]: Ah, ok, that makes sense. Thanks for the explanation!

**Comment by RyanTheOptimist** [comment]
Looks like the failure are real.

https://github.com/envoyproxy/envoy/actions/runs/5791108893/job/15695371848?pr=28883

https://dev.azure.com/cncf/envoy/_build/results?buildId=145421&view=logs&jobId=c95d62a5-6bb9-58f0-2a02-7575fd482a15&j=c95d62a5-6bb9-58f0-2a02-7575fd482a15&t=76c9b3e6-0d40-57e9-06f3-2219e9228a0a

**Review by mattklein123 — APPROVED** [review]

File: api/envoy/config/core/v3/protocol.proto [hunk 4]
```diff
@@ -94,6 +94,12 @@ message QuicProtocolOptions {
   // Probes the peer at the configured interval to solicit traffic, i.e. ACK or PATH_RESPONSE, from the peer to push back connection idle timeout.
   // If absent, use the default keepalive behavior of which a client connection sends PINGs every 15s, and a server connection doesn't do anything.
   QuicKeepAliveSettings connection_keepalive = 5;
+
+  // An array of QUIC connection options represented in a serialized string, separated by commas.
```
Comment (danzh2010) [inline api/envoy/config/core/v3/protocol.proto:98]: nits: A comma-separated list of strings representing QUIC connection options defined in QUICHE  and to be sent by upstream connections.
Can you reference to the connection options definition file in QUICHE?
